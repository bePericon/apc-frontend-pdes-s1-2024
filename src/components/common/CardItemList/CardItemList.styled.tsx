import styled, { css } from "styled-components";

export const StyledCategoryCard = styled.div`
  min-height: 72px;
  display: flex;
  align-items: center;
  padding: 8px;
  border-radius: 8px;
  box-shadow: ${({ theme }) => theme.boxShadows.whiteS};
  background-color: #fff;
  cursor: pointer;

  &:hover {
    background-color: #FAF0CA;
    outline: none;
  }

  &.focused {
    background-color: #fff;
    outline: none;
  }

  ${({ theme }) => {
    return css`
      @media screen and (max-width: ${theme.breakpoints.s}) {
        width: 264px;
        height: auto;
        min-height: 52px;
        padding: 8px;
      }
    `;
  }}
`;

export const StyledCategoryCardInner = styled.div`
  display: flex;
`;

export const StyledButton = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
`;

export const StyledInfo = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 4px;
  min-width: 500px;
  max-width: 500px;
  padding-top: 4px;

  [role="heading"] {
    line-height: 1.43;
    letter-spacing: normal;
  }

  ${({ theme }) => {
    return css`
      @media screen and (max-width: ${theme.breakpoints.m}) {
        min-height: auto;
      }

      @media screen and (max-width: ${theme.breakpoints.s}) {
        min-width: 180px;
      }
    `;
  }}
`;

export const StyledFirstIcon = styled.div`
  align-self: flex-start;
  margin-right: 8px;
`;

export const StyledRenderIcon = styled.div`
  align-self: flex-start;
`;

export const StyledLastIcon = styled.div`
  align-self: center;
`;

export const StyledColumnItems = styled.div`
  display: flex;
  flex-direction: column;

  gap: 24px;
  padding: 20px 0 20px;

  ${({ theme }) => {
    return css`
      @media screen and (max-width: ${theme.breakpoints.m}) {
        flex-direction: column;
        gap: 16px;
      }
    `;
  }}
`;