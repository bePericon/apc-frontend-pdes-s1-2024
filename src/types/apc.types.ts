export interface Role {
  _id: string;
  name: string;
  description: string;
  permissions: [
    {
      _id: string;
      name: string;
      description: string;
      creationDate: string;
    }
  ];
  creationDate: string;
}

export interface Favorite {
  _id: string;
  user: string;
  itemId: string;
  comment: string;
  rating: number;
  creationDate: string;
}

export interface User {
  _id: string;
  name: string;
  surname: string;
  username: string;
  email: string;
  password: string;
  roles: Role[];
  favorites: Favorite[];
  creationDate: string;
}
