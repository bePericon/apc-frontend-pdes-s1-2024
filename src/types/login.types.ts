export interface LoginData {
    email: string
    password: string
}

export interface LoginUserData {
    user: UserData
    // token: string;
}

export interface UserData {
    name: string
    surname: string
    username: string
    email: string
    createdDate?: Date
}

export interface SignUpData extends LoginData {
    name: string
}
