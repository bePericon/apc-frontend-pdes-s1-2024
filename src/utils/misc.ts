

export const numberWithCommas = (x: any) => {
    let parts = x.toString().split('.')
    parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, '.')
    if (parts[1]?.length === 1) {
        // IF THERE IS ONLY ONE DIGIT AFTER THE COMMA, i.e. 25.5, MAKE IT 25.50
        parts[1] = parts[1] + '0'
    }
    return parts.join(',')
}