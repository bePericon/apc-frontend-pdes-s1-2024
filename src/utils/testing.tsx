import React, { PropsWithChildren } from "react";
import { render as rtlRender } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../redux/store";
import { theme } from "@/styles/theme";
import { DefaultTheme, ThemeProvider } from "styled-components";

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

// Wait for the promises to finish
const flushPromises = () => new Promise(process.nextTick);

function render(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>
      <ThemeProvider theme={theme as DefaultTheme}>{children}</ThemeProvider>
    </Provider>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export * from "@testing-library/react";
export { render, flushPromises };
