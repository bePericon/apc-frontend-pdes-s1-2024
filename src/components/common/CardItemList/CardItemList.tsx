import {
  StyledButton,
  StyledCategoryCard,
  StyledFirstIcon,
  StyledInfo,
  StyledLastIcon,
} from "./CardItemList.styled";
import { ChevronRight } from "@mui/icons-material";
import Image from "next/image";
import { Typography } from "@mui/material";
import DialogItemList from "../DialogItemList/DialogItemList";
import { useState } from "react";
import { Product } from "@/types/meli.types";

interface CardItemListProps {
  item: Product;
}

const CardItemList = ({ item }: CardItemListProps) => {
  const [open, setOpen] = useState(false);

  const handleOnClick = async () => setOpen(true);

  const handleOnClose = () => setOpen(false);

  return (
    <>
      <StyledCategoryCard onClick={handleOnClick}>
        <StyledFirstIcon>
          {item.thumbnail && (
            <Image
              src={item.thumbnail}
              width={90}
              height={90}
              alt={item.thumbnail_id}
              style={{ borderRadius: 8 }}
            />
          )}
        </StyledFirstIcon>
        <div>
          <StyledInfo>
            <Typography>{item.title}</Typography>
          </StyledInfo>
        </div>
        <StyledLastIcon>
          <StyledButton>
            <ChevronRight />
          </StyledButton>
        </StyledLastIcon>
      </StyledCategoryCard>
      <DialogItemList onClose={handleOnClose} open={open} item={item} />
    </>
  );
};

export default CardItemList;
