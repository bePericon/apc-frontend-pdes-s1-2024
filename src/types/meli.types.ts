import { User } from './apc.types'

export interface FavoriteProduct {
    favoriteId?: string
    rating?: number
    comment?: string
    isFavorite?: boolean
}

export interface Product extends FavoriteProduct {
    title: string
    thumbnail: string
    thumbnail_id: string
    pictures: any[]
    price: number
    itemId: string

    userId?: string
    createdDate?: string
    hydrated?: HydratedProduct
}

export interface PurchaseProduct extends Product {
    purchaseId: string
    price: number
    quantity: number
}

export interface HydratedProduct {
    itemId: string
    title: string
    thumbnail: string
    thumbnail_id: string
    pictures: Picture[]
    price: number
}

export interface Picture {
    quality: string
    id: string
    url: string
    secure_url: string
    size: string
    max_size: string
}
