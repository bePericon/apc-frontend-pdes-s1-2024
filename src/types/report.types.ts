import { User } from './apc.types'

export interface FavoriteItem {
    _id: string
    user: User
    itemId: string
    comment: string
    rating: number
}

export interface FavoriteHydratedReport  {
    title: string;
    thumbnail: string;
    thumbnail_id: string;
    pictures: any[]
    price: number
}

export interface FavoriteReportItem {
    _id: string // itemId
    items: FavoriteItem[]
    count: number // items length
    averageRating: number
    hydrated: FavoriteHydratedReport // only data from Meli
}
