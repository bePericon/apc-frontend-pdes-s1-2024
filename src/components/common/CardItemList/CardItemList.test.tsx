import { render } from "@/utils/testing";
import CardItemList from "./CardItemList";
import { Product } from "@/types/meli.types";
import "@testing-library/jest-dom";

describe("CardItemList", () => {
  const product: Product = {
    _id: "",
    title: "Product name",
    thumbnail: "",
    thumbnail_id: "",
    pictures: [],
    price: 100,
  };

  it("should render correctly", () => {
    const result = render(<CardItemList item={product} />);

    expect(result).toMatchSnapshot();
  });

  it("should show correct title", () => {
    const { getByText } = render(<CardItemList item={product} />);

    expect(getByText(/Product name/)).toBeInTheDocument();
  });
});
