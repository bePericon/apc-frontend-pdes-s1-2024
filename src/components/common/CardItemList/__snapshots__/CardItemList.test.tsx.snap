// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItemList should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-5d2a921f-0 kRThmr"
      >
        <div
          class="sc-5d2a921f-4 iVmgRH"
        />
        <div>
          <div
            class="sc-5d2a921f-3 fqvml"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Product name
            </p>
          </div>
        </div>
        <div
          class="sc-5d2a921f-6 fYXory"
        >
          <div
            class="sc-5d2a921f-2 hdHNuM"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ChevronRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-5d2a921f-0 kRThmr"
    >
      <div
        class="sc-5d2a921f-4 iVmgRH"
      />
      <div>
        <div
          class="sc-5d2a921f-3 fqvml"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Product name
          </p>
        </div>
      </div>
      <div
        class="sc-5d2a921f-6 fYXory"
      >
        <div
          class="sc-5d2a921f-2 hdHNuM"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
